tools/gfx/persistent-shader-cache.cpp

```diff
@@ -0,0 +1,180 @@
+    if (shaderCacheFileSystem)
+    {
+        shaderCacheFileSystem->queryInterface(ISlangMutableFileSystem::getTypeGuid(), (void**)mutableShaderCacheFileSystem.writeRef());
+    }
+
+    loadCacheIndexFromFile();
+}
+
+// Load a previous cache index saved to disk. If not found, create a new cache index
+// and save it to disk as filename.
+void PersistentShaderCache::loadCacheIndexFromFile()
+{
+    ComPtr<ISlangBlob> indexBlob;
+    if (SLANG_FAILED(shaderCacheFileSystem->loadFile(indexFilename.getBuffer(), indexBlob.writeRef())))
+    {
+        // Cache index not found, so we'll create and save a new one.
+        if (mutableShaderCacheFileSystem)
+        {
+            mutableShaderCacheFileSystem->saveFile(indexFilename.getBuffer(), nullptr, 0);
+            return;
+        }
+        // Cache index not found and we can't save a new one due to the file system being immutable.
+        return;
+    }
+
+    String indexString;
+    File::readAllText(indexFilename, indexString);
+
+    List<UnownedStringSlice> lines;
+    StringUtil::calcLines(indexString.getUnownedSlice(), lines);
+    for (auto line : lines)
+    {
+        List<UnownedStringSlice> digests;
+        StringUtil::split(line, ' ', digests); // This will return our two hashes as two elements in digests.
+        auto dependencyDigest = DigestUtil::fromString(digests[0]);
+        auto astDigest = DigestUtil::fromString(digests[1]);
+
+        ShaderCacheEntry entry = { dependencyDigest, astDigest };
+        auto entryNode = entries.AddLast(entry);
+        keyToEntry.Add(dependencyDigest, entryNode);
+    }
+}
+
+LinkedNode<ShaderCacheEntry>* PersistentShaderCache::findEntry(const slang::Digest& key, ISlangBlob** outCompiledCode)
+{
+    LinkedNode<ShaderCacheEntry>* entryNode;
+    if (!keyToEntry.TryGetValue(key, entryNode))
+    {
+        // The key was not found in the cache, so we return nullptr.
+        *outCompiledCode = nullptr;
+        return nullptr;
+    }
+
+    // If the key is found, we need to move its corresponding entry to the front of
+    // the list and load the stored contents from disk.
+    shaderCacheFileSystem->loadFile(DigestUtil::toString(key).getBuffer(), outCompiledCode);
+    if (entries.FirstNode() != entryNode)
+    {
+        entries.RemoveFromList(entryNode);
+        entries.AddFirst(entryNode);
+        if (mutableShaderCacheFileSystem)
+        {
+            saveCacheIndexToFile();
+        }
+    }
+    return entryNode;
+}
+
+void PersistentShaderCache::addEntry(const slang::Digest& dependencyDigest, const slang::Digest& astDigest, ISlangBlob* compiledCode)
+{
+    if (!mutableShaderCacheFileSystem)
+    {
+        // Should not save new entries if the underlying file system isn't mutable.
+        return;
+    }
+    
+    // Check that we do not exceed the cache's size limit by adding another entry. If so,
+    // remove the least recently used entry first.
+    while (entryCountLimit > 0 && entries.Count() >= entryCountLimit)
```

Comment:
Need to embed this while loop in an `if (!mutableShaderCacheFileSystem)` otherwise you could risk entering infinite loop here. When writing a `while`, always think "could this lead to infinite loop under some situations?"

Reply:
Ah, nevermind, I see you have an early return above, everything is good then.

Reply:
I don't think this will ever occur because of the `if (!mutableShaderCacheFileSystem)` check at the very start. The only way this infinitely loops without crashing is if `deleteLRUEntry` immediately returns, but that only occurs when the file system isn't mutable. We should never make it to this loop if that's the case.